### DISCUSSION TITLE
Preprocessing docstrings (for link regex-replace)

### DISCUSSION THREAD

**Discussion by ssweber** [mkdocstrings/mkdocstrings#596] [Q&A] [answered]
I wasn’t quite sure if this should be implemented in mkdocstrings, autorefs, or griffe.

I have a project that uses a simple convention for referencing other classes/functions/attributes:

Example:
```
See `ClassName.FunctionName` documentation for more details
```

Is there a way for me to hook a function to replace:

```python
`ClassName.fn` with [fn][long_prefix.ClassName.fn]

or

`ClassName` with [ClassName][long_prefix.ClassName]
```

Before mkdocs/autorefs converts it into a link?

For now I’ll run a pre-processing script on the throw-away branch before running mkdocs serve/gh-deploy.

I think this allows a simpler and flexible solution to others desire for automatic links (See Also section), and/or relative link handling… without having to adding a bunch of markdown to the source-code.

**Comment by pawamoy** [comment, marked as answer]
You could pre-process docstrings with a Griffe extension. See the docs: https://mkdocstrings.github.io/griffe/extensions/.
Running a regex-replace on the docstring could do the trick, though it would be your responsibility to find out the complete path of the object. That seems doable given you'll know the parents of the object being pre-processed.

**Reply by ssweber** [reply]
Thanks! I always forget about GitHub’s “Discussion” section. I’ll report back with a gist once I have it working so others can benefit. 😄

**Reply by pawamoy** [reply]
Great, don't hesitate to ask for help if the docs are unclear :slightly_smiling_face:

**Reply by ssweber** [reply]
[Simple replacement for monolithic module](https://gist.github.com/ssweber/a6ea72fb5b10f6d4d91d14367fba9f9e)

This works for me, although it would need to be made smarter for others that have multiple modules. It's already helping me catch documentation that has invalid references :)

Quick question: What is your approach to documenting and creating API links dataclass attributes that are set in the __post_init__? Is there anyway I can get mkdocstrings to create those like cls_var and var1 in this example?

See this example:
```python
class TestClass:
    """This is the TestClass docstring"""
    
    cls_var: ClassVar[int] = 1
    """This is the cls_var docstring"""
    
    var1: int
    """var docstring"""
    
    def __post_init__(self):
        self.var2 = 2
        """var2 docstring"""

class TestLinks:
    """Links to `TestClass`, `TestClass.cls_var`, `TestClass.var1`, `TestClass.var2`"""
```

with my RegexUrl makes:

![image](https://github.com/mkdocstrings/mkdocstrings/assets/57631333/a5e4afc6-df20-48cd-9eb7-66920ae06622)

**Reply by ssweber** [reply]
I tried adding a docstring to __post_init__, but that still didn’t make var2 to show up.

**Reply by ssweber** [reply]
Perhaps that’s simply unsupported at the moment? 

>Support for documented attributes: attributes (variables) followed by a docstring (triple-quoted string) will be recognized by Griffe in modules, classes and even in __init__ methods.

If it’s not too difficult, could you point me in the right direction where I could work on adding __post_init__ to this list?

**Reply by ssweber** [reply]
Looks like perhaps in https://github.com/mkdocstrings/griffe/blob/1e7b869b7345b30185e6bf634bc4f848c940172f/src/griffe/agents/visitor.py

**Reply by ssweber** [reply]
Sorry for the noise, starting to wrap my head around how the extensions work in griffe (really nice btw!).

Looks like I can for the time-being hook into  on_attribute_node, copying over how it grabs the attribute docstring for `__init__` and make it work for `__post_init__`. May need to strip the “self.” from the name? I’ll add another gist once I’ve figured it out and (if you like the idea), create a pull request to add to main.

**Reply by pawamoy** [reply]
No worries, I work this way too: creating issues/discussions and answering myself haha. Will try to answer later.

**Reply by ssweber** [reply]
That’s funny :) I think it’s helpful for the next person who has a similar issue. Thanks for being welcoming.

**Comment by ssweber** [comment]
I opened a small issue/pr in the griffe branch to handle `__post_init__`. I didn't know how to easily write an extension to do it externally.

**Reply by pawamoy** [reply]
Yep, to support this in extension we would probably need to expose the visitor/inspector to extensions.
I answered in the issue you opened :slightly_smiling_face:
